src/modules/commander/Commander.cpp

```diff
@@ -497,6 +497,134 @@ transition_result_t Commander::disarm(arm_disarm_reason_t calling_reason)
 	return arming_res;
 }
 
+transition_result_t
+Commander::try_mode_change(main_state_t desired_mode, const bool enable_fallback, const bool notify_user)
+{
+	transition_result_t res = main_state_transition(_status, desired_mode, _status_flags, &_internal_state);
+
+	if (res == TRANSITION_DENIED) {
+
+		if (desired_mode == commander_state_s::MAIN_STATE_OFFBOARD) {
+			/* offboard does not have a fallback */
+			if (notify_user) {
+				print_reject_mode("Offboard");
+			}
+
+			return res;
+		}
+
+		if (desired_mode == commander_state_s::MAIN_STATE_AUTO_MISSION) {
+			/* fall back to loiter */
+			if (notify_user) {
+				print_reject_mode("Auto Mission");
+			}
+
+			if (enable_fallback) {
+				desired_mode = commander_state_s::MAIN_STATE_AUTO_LOITER;
+				res = main_state_transition(_status, desired_mode, _status_flags, &_internal_state);
+			}
+		}
+
+		if (desired_mode == commander_state_s::MAIN_STATE_AUTO_RTL && (res == TRANSITION_DENIED)) {
+			/* fall back to loiter */
+			if (notify_user) {
+				print_reject_mode("Auto RTL");
+			}
+
+			if (enable_fallback) {
+				desired_mode = commander_state_s::MAIN_STATE_AUTO_LOITER;
+				res = main_state_transition(_status, desired_mode, _status_flags, &_internal_state);
+			}
+		}
+
+		if (desired_mode == commander_state_s::MAIN_STATE_AUTO_LAND && (res == TRANSITION_DENIED)) {
```

Comment:
Is the additional check for `(res == TRANSITION_DENIED)` to make sure we don't get there through some previous fallback? If so, then we should probably use an `else if` structure to prevent that overall.

Reply:
@MaEtUgR explained it, makes sense.